Allow beam sync to start off from a trusted checkpoint

cburgdorf opened the pull request:
### What was wrong?

Relates to #892 

Before beam sync can start of doing its "beam magic", it still needs to download the chain of headers. While that's not the end of the world, it would be much nicer if the user could provide a checkpoint for a header that beam sync would trust (to start with) and build up on top.

This allows beam sync to get to work *immediately* and fill up missing headers (and eventually also blocks) in the background.

### How was it fixed?

1. Add a `SyncLaunchStrategyAPI` with two implementations:

-  created a new class `Checkpoint` that represents a blockhash, a blocknumber as well as a score
- `FromGenesisLaunchStrategy`: Download the full header chain first (same as today)
- `FromCheckpointLaunchStrategy`: Launch from a given checkpoint

2. Added a new parameter `--beam-from-checkpoint` that takes an URI as a form of micro DSL.

This is following what @pipermerriam proposed in #892 but it's been a bit controversial and I'm open to other ideas. However, here's my attempt to defend it:

For the manual checkpoints (e.g. not retrieving one through the etherscan API) there are three parts to a checkpoint, the blockhash, the blocknumber and the score. Creating three different command line parameters for this:

- A. Is hard to remember (I think it's easier to get used to some form of micro DSL)
- B. Makes Trinitys invocation command much longer
- C. Rips apart three things that essentially need to be supplied together

### Seeing it in action:

1. `rm -rf /home/cburgdorf/.local/share/trinity/mainnet/chain-eth1/`

2. `trinity --sync-mode beam  --beam-from-checkpoint="eth://block/byhash/0x113f05289c685eb5b87d433c3e09ec2bfa51d6472cc37108d03b0113b11e3080?number=8360824&score=11,487,662,456,884,849,810,705"` 

Then observe:

```
INFO  08-16 16:54:26       BeamSyncService  Starting beam-sync; using checkpoint: <Checkpoint block hash=#113f..3080 block number=8360824 score=11487662456884849810705>
...
INFO  08-16 16:54:59     HeaderOnlyPersist  Imported 192 headers in 0.21 seconds, new head: <BlockHeader #8361016 b7e740c2>
    INFO  08-16 16:55:01     HeaderOnlyPersist  Imported 388 headers in 0.06 seconds, new head: <BlockHeader #8361404 44c602c9>
    INFO  08-16 16:55:02     HeaderOnlyPersist  Caught up to skeleton peer. Switching to beam mode at <BlockHeader #8362170 e4e1ea7c>
    INFO  08-16 16:55:02            BeamSyncer  Downloading 8 block bodies for uncle validation, before <BlockHeader #8362170 e4e1ea7c>
    INFO  08-16 16:55:02            BeamSyncer  Getting recent block data for uncle validation
    INFO  08-16 16:55:06            BeamSyncer  Have all data needed for Beam validation, continuing...
    INFO  08-16 16:55:06  HeaderCheckpointSyncer  Choosing (<eth.rlp.headers.BlockHeader object at 0x7f6c8effca90>,) as checkpoint headers to sync from
    INFO  08-16 16:55:06        BeamDownloader  Starting beam state sync
    INFO  08-16 16:55:06        BeamDownloader  Beam-Sync: all=0  urgent=0  pred=0  all/sec=0  urgent/sec=0  reqs=0  pred_reqs=0  timeouts=0
    INFO  08-16 16:55:06           ETHPeerPool  Adding ETHPeer <Node(0x1fdd@91.195.52.166)> to pool
    INFO  08-16 16:55:06  RegularChainBodySyncer  Waited 0.6s for <BlockHeader #8362170 e4e1ea7c> body
    INFO  08-16 16:55:06     BeamBlockImporter  Beam importing <BlockHeader #8362170 e4e1ea7c> (69 txns) ...
    INFO  08-16 16:55:16        BeamDownloader  Beam-Sync: all=959  urgent=758  pred=201  all/sec=95  urgent/sec=75  reqs=44  pred_reqs=0  timeouts=1
    INFO  08-16 16:55:23           ETHPeerPool  Adding ETHPeer <Node(0x58dc@23.111.173.234)> to pool
    INFO  08-16 16:55:26        BeamDownloader  Beam-Sync: all=2943  urgent=800  pred=2143  all/sec=145  urgent/sec=39  reqs=86  pred_reqs=0  timeouts=1
    INFO  08-16 16:55:26           ETHPeerPool  Connected peers: 0 inbound, 7 outbound
    INFO  08-16 16:55:36        BeamDownloader  Beam-Sync: all=4693  urgent=838  pred=3855  all/sec=155  urgent/sec=27  reqs=124  pred_reqs=0  timeouts=1
    INFO  08-16 16:55:46        BeamDownloader  Beam-Sync: all=5761  urgent=890  pred=4871  all/sec=143  urgent/sec=22  reqs=176  pred_reqs=0  timeouts=1
    INFO  08-16 16:55:56        BeamDownloader  Beam-Sync: all=8190  urgent=926  pred=7264  all/sec=163  urgent/sec=18  reqs=216  pred_reqs=4  timeouts=1
    INFO  08-16 16:56:06        BeamDownloader  Beam-Sync: all=10329  urgent=967  pred=9362  all/sec=171  urgent/sec=16  reqs=257  pred_reqs=4  timeouts=1
    INFO  08-16 16:56:16        BeamDownloader  Beam-Sync: all=11746  urgent=1010  pred=10736  all/sec=167  urgent/sec=14  reqs=300  pred_reqs=4  timeouts=1
    INFO  08-16 16:56:26        BeamDownloader  Beam-Sync: all=12933  urgent=1051  pred=11882  all/sec=161  urgent/sec=13  reqs=341  pred_reqs=4  timeouts=1
    INFO  08-16 16:56:26           ETHPeerPool  Connected peers: 0 inbound, 7 outbound
    INFO  08-16 16:56:36        BeamDownloader  Beam-Sync: all=14168  urgent=1098  pred=13070  all/sec=156  urgent/sec=12  reqs=388  pred_reqs=4  timeouts=1
    INFO  08-16 16:56:46        BeamDownloader  Beam-Sync: all=15599  urgent=1137  pred=14462  all/sec=155  urgent/sec=11  reqs=427  pred_reqs=4  timeouts=1
    INFO  08-16 16:56:53           ETHPeerPool  Adding ETHPeer <Node(0xf9ae@203.238.15.78)> to pool
    INFO  08-16 16:56:56           ETHPeerPool  Adding ETHPeer <Node(0x0122@47.254.45.126)> to pool
    INFO  08-16 16:56:56           ETHPeerPool  ETHPeer <Node(0x0122@47.254.45.126)> finished[DisconnectReason.subprotocol_error], removing from pool
    INFO  08-16 16:56:56        BeamDownloader  Beam-Sync: all=17160  urgent=1179  pred=15981  all/sec=155  urgent/sec=10  reqs=469  pred_reqs=4  timeouts=1
    INFO  08-16 16:57:01  RegularChainBodySyncer  Imported block 8362170 (69 txs) in 114.37 seconds, with 2m7s lag


```

### To-Do

- [x] Hook it up to the command line (currently uses a hardcoded checkpoint)
- [x] polish local py-evm changes and open up Py-EVM PR
- [x] print out used checkpoint
- [x] switch to proper Py-EVM release
- [x] parse out seperators from block number / score
- [x] Get rid of FIXMEs
- [x] add a test that launches sync from a checkpoint
- [x] Clean up commit history
- [x] Add entry to the [release notes]

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](http://www.awesomelycute.com/gallery/2014/02/cutest-wild-animals-2944.jpg)

cburgdorf commented on setup.py:
```diff
@@ -3,7 +3,7 @@
 from setuptools import setup, find_packages
 
-PYEVM_DEPENDENCY = "py-evm==0.3.0a3"
+PYEVM_DEPENDENCY = "py-evm@git+https://github.com/cburgdorf/py-evm.git@christoph/spike/persist-orphan-header"  # noqa: E501
```
Can't merge this before we cut a new py-evm release

cburgdorf commented:
@pipermerriam @carver This isn't ready to be merged yet because we need to cut a new Py-EVM release first. It also has no tests yet (That said, at least the current tests prove that it's not breaking existing functionality)

But I'm curious about your feedback on the taken approach and maybe things that are missing.

cburgdorf commented on trinity/sync/common/headers.py:
```diff
@@ -75,6 +81,93 @@
+            # in the usual way.
+
+            # TODO: Do we need to do any further validation before we persist this?
+            await self._db.coro_persist_checkpoint_header(header, self._checkpoint.score)
```
@pipermerriam @carver 

I wonder if we would need to perform any special kind of validation here first. We essentially only the hash and then save it in our db with our given score. What if a peer serves us a header with that hash that is actually pure garbage (only the hash field is set to match our expectations but the rest doesn't match).

cburgdorf replied:
Also, as the comment hints, I'm not happy about the side effect of this method but I'm also unsure about better alternatives.

carver replied:
> What if a peer serves us a header with that hash that is actually pure garbage (only the hash field is set to match our expectations but the rest doesn't match).

The header hash isn't a field in the header RLP returned by the client, we calculate it locally. Also, we verify that the first header received over the network has the hash that we requested in `BaseBlockHeadersValidator`. So I think we're doing okay on validation that the header has the specified hash.

cburgdorf replied:
>The header hash isn't a field in the header RLP returned by the client, we calculate it locally

:see_no_evil: SURE!

See, I'm a real blockchain professional. Been in this space for 30 years :joy:

carver commented:
Looks/sounds like this is a WIP, so just adding it to the title.

cburgdorf commented:
@pipermerriam @carver This is ready for a full review. I also added a test that syncs from a checkpoint (and by which I also fixed a bug that existed in a previous version of this PR).

The only things missing from this PR (afaik) is support for directly fetching checkpoints through etherscan or infura which I intend to add with another PR.

cburgdorf commented on tests/core/p2p-proto/test_sync.py:
```diff
@@ -206,7 +235,7 @@ def small_header_batches(monkeypatch):
         target_block_number = min(beam_to_block + 10, 129)
         target_head = chaindb_churner.get_canonical_block_header_by_number(target_block_number)
-        await wait_for_head(chaindb_fresh, target_head, sync_timeout=4)
+        await wait_for_head(chaindb_fresh, target_head, sync_timeout=5)
```
I needed this locally to work with these tests but I *think* I've also seen this timeout in CI from time to time so maybe increasing this will make this test less flaky under CI, too.

pipermerriam commented on trinity/plugins/builtin/syncer/cli.py:
```diff
@@ -0,0 +1,73 @@
+    return Checkpoint(Hash32(decode_hex(block_hash)), BlockNumber(int(block_number)), int(score))
+
+
+class NormalizeCheckpointUri(argparse.Action):
```
*naming nitpick*.  I'd all-caps `URI`

pipermerriam commented on trinity/plugins/builtin/syncer/plugin.py:
```diff
@@ -200,6 +201,13 @@ def configure_parser(cls, arg_parser: ArgumentParser) -> None:
+        arg_parser.add_argument(
+            '--beam-from-checkpoint',
+            action=NormalizeCheckpointUri,
+            help="Start beam sync from a trusted checkpoint",
```
Help text here could include the template for how a checkpoint URI should be formatted.

carver commented on trinity/sync/common/headers.py:
```diff
@@ -75,6 +81,92 @@
+
+    async def get_starting_block_number(self) -> BlockNumber:
+        default_starting_block_number = await self._genesis_strategy.get_starting_block_number()
+        if default_starting_block_number < self._checkpoint.block_number:
```
Okay, I see how the block number is used... I still think it's worth adding a pre-sync stage to download just the one checkpoint header that was specified. That will give a nice clear error message if the checkpoint was mis-specified somehow (rather than a surprising/confusing skeleton sync one).

So the rough idea is:
- loop over peers, asking for the header, until you get it back
- block any other syncing until ^ is complete
- maybe timeout after 60s (given that you are actually connected to a peer), and print a clear log like f"your connected peers did not respond with the checkpoint header specified in the `--whatever` argument: {checkpoint}"

Maybe this belongs in something like a new `await SyncLaunchStrategyAPI.fulfill_prerequisites()` (probably with a better name). The genesis one could be a noop, and the other one could explicitly download the header. An awkward part is that it would need the peer pool or something as an argument. But maybe that's reasonable. :man_shrugging: 

---

I just noticed that maybe it removes the side-effect mentioned here, too:

```py
            # The side effect is unfortunate but it's a special case of this strategy that we
            # have to persist this header at this point. After this, the sync will continue
            # in the usual way.
```

cburgdorf replied:
How about we delay that to the next PR that will add the functionality to defer to *etherscan | infura* to figure out the specific checkpoint?

With these, we'll first need to resolve to a checkpoint anyway before the sync starts so it is somewhat related.

I'm leaning towards the idea that this would happen outside/before the `SyncLaunchStrategyAPI` comes into play but that's just a feeling at this point. But I do think that postponing this until we tackle the *etherscan | infura* resolving makes sense as there may be case for a streamlined API.

In fact, instead of asking our peers for the block (and therefore depending on the peer pool) we *could* also default to the etherscan API to resolve information that wasn't specified by the user (in fact, that would then also work for `score`).

Well, these are just loose thoughts, my main point is, maybe we first tackle the *etherscan | infura* resolving to see how that turns out / forms the API.

This would leave us with the options to:

- merge this already (addressed all other poiints)
- keep it open and add *etherscan | infura* resolving to this PR

carver replied:
Hm, I'm not wild about using etherscan/etc for anything other than giving us a checkpoint hash & score. Open to the conversation, though.

I don't necessarily think that how we resolve the block number requires us to bring in 3rd-party services.

The main reason to tackle this now is: we can entirely remove the block number CLI specification, and logic handling, from this PR now. Otherwise, we are adding it as something that we expect to become cruft soon.

cburgdorf replied:
My initial thought was that it would probably be faster to resolve the block number through 3rd party services because we don't need to wait for connected peers.

But now that I give it more thought, maybe that doesn't really matter because having peers is a pre-condition to syncing anyway so by the time that we'd be ready to start syncing, we'll also be ready to resolve the given hash with a peer.

I'm gonna give it a try.

carver commented on trinity/sync/common/headers.py:
```diff
@@ -75,6 +81,92 @@
+        ...
+
+    @abstractmethod
+    async def get_expected_score(self) -> int:
```
The word `expected` in here was a surprise to me. Isn't this just the current head's local score? `get_current_score()` or `get_head_score()`, maybe?

Either way, it sounds like there is some subtlety in here that is probably worth adding a docstring for.

cburgdorf replied:
It's not necessary the score of the head because it may also be the score of the defined checkpoint. What about `get_launch_score` or `get_start_score`?

carver replied:
I am thinking of setting the checkpoint as like the user pre-defining a starting head, like setting a genesis file is defining a starting head. Are you shy to call it the head because we haven't verified it back to the genesis yet?

If we called `chain.get_score(chain.get_canonical_head())`, then I believe it will return the same thing an `get_expected_score()`, whether or not we used a checkpoint. So `get_head_score()` still makes a lot of sense to me.

`launch` and `start` scores both sound like things that wouldn't/shouldn't change over time as you sync, but this returned value will.

cburgdorf replied:
Yes, I think you are right. Storing the header before sync even starts seems to eliminate some of the different code paths needed within the syncing logic. That's a good thing. I'm gonna give it a try.

carver commented on trinity/sync/common/headers.py:
```diff
@@ -75,6 +81,92 @@
 )
 
 
+class SyncLaunchStrategyAPI(ABC):
```
This API neatly peels out a contained piece of functionality. Nice! :+1:

carver commented on trinity/sync/common/headers.py:
```diff
@@ -75,6 +81,92 @@
+        ...
+
+
+class FromGenesisLaunchStrategy(SyncLaunchStrategyAPI):
```
Maybe these belong in a new `sync/common/strategies.py`, since this module is already pretty big.

cburgdorf replied:
Yep, agreed!

carver commented on trinity/sync/common/headers.py:
```diff
@@ -243,17 +340,19 @@ def __init__(self,
+        )
 
-        if not first_is_present:
+        if not can_start_with_header:
```
Nice generalization of the existing code. :+1:

carver commented on trinity/sync/common/headers.py:
```diff
@@ -243,17 +340,19 @@ def __init__(self,
             for parent, child in sliding_window(2, skeleton_launch_headers):
-                is_present = await self.wait(self._db.coro_header_exists(child.hash))
-                if not is_present:
+                usable_header = await self.wait(self._launch_strategy.can_start_with_header(child))
```
Maybe `is_usable_header`? Although, maybe this whole line change reverts, if we do the pre-downloading of the checkpoint header.

cburgdorf commented on newsfragments/921.feature.rst:
```diff
@@ -0,0 +1,8 @@
+Allow beam sync to start from a trusted checkpoint.
+Specify a checkpoint via CLI parameter such as:
+
+--beam-from-checkpoint="block://blockhash/blocknumber/score"
```
I like your proposed URI format. Let's do that :+1:

pipermerriam replied:
I want to push back and propose we stick with the existing standard from [BIP-122](https://github.com/bitcoin/bips/blob/master/bip-0122.mediawiki)

cburgdorf replied:
>I want to push back and propose we stick with the existing standard from BIP-122

If I understand that standard correctly it doesn't cover how to retrieve a block hash **and** a score together (or even block hash + score + number if we want it self contained, independent from requesting out to get the block number from somewhere else).

I think @carver brought up good reasons to adjust the theme (and the PR implements it already).

Basically the only thing I personally care about is to have a scheme with a command line parameter because I feel multiple hurt the UX.

How do we move on here?

pipermerriam replied:
Ah, I realize this is missing `score`.... hrm,  I don't have a good answer and don't want to hold this up.  carry on with whatever you were doing before and we can adjust the scheme later if we find something we really like.

cburgdorf commented on tests/core/p2p-proto/test_sync.py:
```diff
@@ -140,14 +167,15 @@ def small_header_batches(monkeypatch):
 # range(1, 130) for beam_to_block. (and optionally follow the instructions at target_head)
 @pytest.mark.asyncio
-@pytest.mark.parametrize('beam_to_block', [1, 66, 68, 129])
+@pytest.mark.parametrize('beam_to_block', [66, 1, 66, 68, 129])
```
Oops, it is!

cburgdorf commented on trinity/sync/beam/service.py:
```diff
@@ -30,17 +31,22 @@ def __init__(
-        head = await self.wait(self.chaindb.coro_get_canonical_head())
-        self.logger.info("Starting beam-sync; current head: %s", head)
+        if self.checkpoint is not None:
+            self.logger.info("Starting beam-sync; using checkpoint: %s", self.checkpoint)
```
Good point :+1:

carver replied:
Seems a little verbose to show `Starting beam-sync` twice, if there is a checkpoint.

cburgdorf replied:
:+1: fixed

cburgdorf commented:
Thank you both for the reviews! I think I addressed all points except one which I explain in this comment.

https://github.com/ethereum/trinity/pull/921#discussion_r315621873

Basically, my thought is to first add the *etherscan | infura* resolving to see how that might form the API.

I can start working on that right away and we could do one of these things:

- keep this PR open until *etherscan | infura* resolving is finished
- merge it already

My preference would be to merge this already but if you prefer not to then that's ok, too :)

cburgdorf commented on trinity/plugins/builtin/syncer/plugin.py:
```diff
@@ -215,6 +226,7 @@ def configure_parser(cls, arg_parser: ArgumentParser) -> None:
             base_db,
             cast(ETHPeerPool, peer_pool),
             event_bus,
+            args.beam_from_checkpoint,
```
@carver I *think* that we could handle the resolving of checkpoints nicely on top of this:



```python

if args.beam_from_checkpoint:
    checkpoint = await checkpoint_resolver.resolve(args.beam_from_checkpoint)

syncer = BeamSyncService(
            chain,
            AsyncChainDB(base_db),
            checkpoint
            ...
)
``` 

This would mean that:

1. `args.beam_from_checkpoint` is just the URI string that was validated by the parser, not the checkpoint itself. Actually, I'd thought this could just be a `Future[Checkpoint]` but that doesn't work because it needs to be pickable. Not big deal.

2. There's no extra resolving complexity *within* the syncing primitives

3. Resolving could theoretically call out to various APIs (peers, infura, etherscan) and just use whoever responds fastest

4. Naturally, the sync won't until the checkpoint is resolved and hence there's nothing special to do *within* `BeamSyncService` to wait on anything.

What do you think?

carver replied:
Yeah, totally, handling it before sync even starts sounds great. I assume this would not only resolve, but also *persist* the header.

I think it's worth building the peer-based resolver now, and removing the block number from the checkpoint. Then some of the follow-on logic gets more unified, when we can assume the starting header (genesis or checkpoint) is already persisted and considered canonical.

cburgdorf commented on trinity/sync/common/strategies.py:
```diff
@@ -0,0 +1,136 @@
+        self._peer_pool = peer_pool
+
+    async def fulfill_prerequisites(self) -> None:
+        max_attempts = 1000
```
@carver instead of a fixed timeout I used a limit of attempts here. The actual time it waits is therefore a bit more dynamic depending on the fact whether we are connected to peers or not. This feels more natural to me. E.g. if I'm in general trouble to find peers it seems reasonable that this task would factor that in and also wait a little longer. Does that make sense?

carver replied:
Yup, sounds good!

cburgdorf commented:
@carver I refactored this again and (to me) it feels indeed much cleaner now :)

It does depend on a new upcoming py-evm version though https://github.com/ethereum/py-evm/pull/1830

(I have to rush to my daughters birthday so skipping further explanations for now)

carver commented on trinity/sync/common/strategies.py:
```diff
@@ -0,0 +1,136 @@
+
+            if not headers:
+                await peer.disconnect(DisconnectReason.useless_peer)
+            elif headers[0].hash != self._checkpoint.block_hash:
```
This is already validated internally to the `get_block_headers()` call, but I guess no objection to re-validating it here, since it's cheap and only happens once.

carver commented on trinity/sync/common/strategies.py:
```diff
@@ -0,0 +1,136 @@
+                await self._db.coro_persist_checkpoint_header(headers[0], self._checkpoint.score)
+                return
+
+            await asyncio.sleep(0.5)
```
Not sure we even need to wait this long, though I'm also not sure it matters much. We can only ask 2 peers per second for the data, but in practice, it should only take a couple tries at most, I hope/expect.

cburgdorf replied:
:+1:  Changed it to `0.05`

carver commented on trinity/sync/common/strategies.py:
```diff
@@ -0,0 +1,136 @@
+        if head.hash == self._checkpoint.block_hash:
+            return head.block_number
+        else:
+            return await self._genesis_strategy.get_starting_block_number()
```
I notice this is making two `coro_get_canonical_head` calls on every check that isn't the checkpoint. Not a big deal, but maybe an easy optimization.

This whole method has exactly the same effect as the genesis launch but with `GENESIS_BLOCK_NUMBER` set to the checkpoint block number. Maybe the sync strategy can collect and cache that value, and then reuse the logic from the genesis version, with a custom minimum block number.

cburgdorf replied:
Ah wait, if I just define `min_block_number = BlockNumber(0)` in the strategy and then implement 

```python
    async def get_starting_block_number(self) -> BlockNumber:
        block_number = await self._genesis_strategy.get_starting_block_number()
        return block_number if block_number > self.min_block_number else self.min_block_number
```

There there is no exception to check, no duplicate checking of `coro_get_canonical_head` and no way we could return a block number that is lower than the one coming from our checkpoint.

Sweet!

carver commented on setup.py:
```diff
@@ -3,7 +3,7 @@
 from setuptools import setup, find_packages
 
-PYEVM_DEPENDENCY = "py-evm==0.3.0a3"
+PYEVM_DEPENDENCY = "py-evm@git+https://github.com/cburgdorf/py-evm.git@ce0470801426166d28b826bbc71a57598ef395d1"  # noqa: E501
```
Just a reminder to get this updated, as you already mentioned.

cburgdorf commented:
Landed :tada: Thanks for guiding me through it @carver !

carver commented on setup.py:
```diff
@@ -3,7 +3,7 @@
 from setuptools import setup, find_packages
 
-PYEVM_DEPENDENCY = "py-evm==0.3.0a3"
+PYEVM_DEPENDENCY = "py-evm==0.3.0a5"  # noqa: E501
```
Hm, I'm tempted to add `yesqa` as a lint check: https://github.com/asottile/yesqa

cburgdorf replied:
Oops, I send a follow up to fix this! And yep, linting for unnecessary `# noqa` comments would be really good!

carver commented on trinity/sync/common/strategies.py:
```diff
@@ -0,0 +1,136 @@
+
+    async def get_starting_block_number(self) -> BlockNumber:
+        block_number = await self._genesis_strategy.get_starting_block_number()
+        return block_number if block_number > self.min_block_number else self.min_block_number
```
IIUC, I think it might be a *good* idea to fail here if you incorrectly call this before finishing the fulfill call. It might help catch a bug earlier.
